### PR TITLE
Feat/exceptionlogging

### PR DESCRIPTION
**Pull request by maqboolayaz**
Logs Creation - Using Spring AOP to provide universal logging of all the exceptions and method calls.

### CODE REVIEW

**Comment by maqboolayaz** [comment]
[ida-ws-dataset.log](https://github.com/dice-group/ida/files/2605375/ida-ws-dataset.log)
[ida-ws-errors.log](https://github.com/dice-group/ida/files/2605376/ida-ws-errors.log)
[ida-ws-requests.log](https://github.com/dice-group/ida/files/2605377/ida-ws-requests.log)

File: ida-ws/src/main/java/upb/ida/logging/AspectLogger.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+           log.info(logMessage.toString());
+           log.setLevel(Level.TRACE);
+           log.trace(returnBean.toString());
+       }
+       else
+       {
+    	   logMessage.append(retVal.toString());
+    	   log.setLevel(Level.INFO);
+    	   log.info(logMessage.toString());
+       }
```
[the remark this answers is not in the document]
Comment (maqboolayaz) [reply ida-ws/src/main/java/upb/ida/logging/AspectLogger.java]: Actually, it's not. the logic is different in both. In the 'if' it is logging the response statement for the Dataset in req/res logs and in the else, it is just sending this whole response to Dataset logs.

**Comment by maqboolayaz** [comment]
Hi @nikit91 , Regarding the future logs for  ChatLog, NlpProcessLog etc. They will all fall in the Trace category too as the dataset response one and will be filtered on the basis of returnBean.getActnCode(). So this will not be an issue in the future. The main logs that must be differentiated in our case here are info, error and trace.

**Comment by nikit91** [comment]
@maqboolayaz,

The log levels are used to define the granularity at which the logs should be maintained for the current instance of the application. For example: RequestResponseLogger might log more things than it is doing currently but at debug or trace level. The loggers are not going to be used only through AoP. Developers might want to log their own stuff which could not be recorded by the AspectLogger.

The same goes with future loggers that we plan, NlpProcessLog might want to log things at different granularity for itself. Some logs related to nlp might be at INFO level and some might be DEBUG or even ERROR. 

Depending upon the which kind of deployment it is, we might want to change the granularity levels of our logs. For example: In DatasetLogger we will log the Metadata about the dataset at INFO level and the actual dataset at DEBUG level. In a production environment we will probably configure the DatasetLogger to only record the info level logs (metadata).

Also, we do not simply plan to keep logging only the "Before" and "After". Many things that are happening during the execution of the methods(custom logs) will be logged as well.

File: ida-ws/src/main/java/upb/ida/logging/AspectLogger.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+       if (args.length > 0) {
+           logMessage.deleteCharAt(logMessage.length() - 1);
+       }
+       logMessage.append(")");
+       logMessage.append(" return Value :");
+       ResponseBean returnBean = (ResponseBean) retVal;
+       if(returnBean.getActnCode() == IDALiteral.UIA_DTTABLE)
+       {
+    	   logMessage.append("Dataset");
+           requestReponseLogger.info(logMessage.toString());
```
Comment (nikit91) [inline ida-ws/src/main/java/upb/ida/logging/AspectLogger.java]: this line is in both if and else, maybe you can put it out

File: ida-ws/src/main/java/upb/ida/logging/AspectLogger.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+@Aspect
+public class AspectLogger {
+
+	@Autowired(required=true)
+	private HttpServletRequest request;
+	@Autowired
+	private ResponseBean response;
+	private Logger requestReponseLogger;
+	private Logger exceptionLogger;
+	private Logger datasetResponseLogger;
```
Comment (nikit91) [inline ida-ws/src/main/java/upb/ida/logging/AspectLogger.java]: You can create singleton spring beans for these loggers and autowire them here instead.

(Ref: https://github.com/dice-group/ida/blob/master/ida-ws/src/main/java/upb/ida/provider/DataDumpProvider.java)

File: ida-ws/log4j.properties [hunk 4]
```diff
@@ -0,0 +1,27 @@
+log4j.logger.RequestResponseLogger=TRACE, reqresappender
+log4j.logger.ExceptionLogger=TRACE, exceptionappender
```
Comment (nikit91) [inline ida-ws/log4j.properties]: this file should be inside src/main/resources

File: ida-ws/src/main/java/upb/ida/logging/AspectLogger.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	
+    /**
+     * Constructor of AspectLogger class
+     */
+	AspectLogger()
+	{
+		BasicConfigurator.configure();
+		String log4jConfigFile = System.getProperty("user.dir") + File.separator + "log4j.properties";
```
Comment (nikit91) [inline ida-ws/src/main/java/upb/ida/logging/AspectLogger.java]: After moving the log4j file to src/main/resources, use FilteUtil.fetchSysFilePath instead.

(Ref: https://github.com/dice-group/ida/blob/master/ida-ws/src/main/java/upb/ida/util/FileUtil.java#L199)

File: ida-ws/src/main/java/upb/ida/logging/AspectLogger.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+		PropertyConfigurator.configure(log4jConfigFile);
+		requestReponseLogger = Logger.getLogger("RequestResponseLogger");
+		exceptionLogger = Logger.getLogger("ExceptionLogger");
+		datasetResponseLogger = Logger.getLogger("DatasetResponseLogger");
+	}
+	
+	@Pointcut("execution(* upb.ida.rest.MessageRestController*.*(..))")
+	public void controller() {
```
Comment (nikit91) [inline ida-ws/src/main/java/upb/ida/logging/AspectLogger.java]: add a comment inside the method body for codacy's quality review to pass

File: ida-ws/src/main/java/upb/ida/logging/AspectLogger.java (hunk 2 above)
Comment (maqboolayaz) [reply ida-ws/src/main/java/upb/ida/logging/AspectLogger.java]: the logic is different in there.

Comment (nikit91) [reply ida-ws/src/main/java/upb/ida/logging/AspectLogger.java]: explain how it's different if I just put the `requestReponseLogger.info(logMessage.toString());` out the if and else block and at the end?

Comment (maqboolayaz) [reply ida-ws/src/main/java/upb/ida/logging/AspectLogger.java]: Actually in the if-statement, it is logging the log message + "Dataset" to the requestresponse log file and the response bean to the dataset log file. While in the else-statement for the normal flow, it is logging the log message + return bean in the requestresponse log file.

File: ida-ws/src/main/java/upb/ida/logging/AspectLogger.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (nikit91) [reply ida-ws/src/main/java/upb/ida/logging/AspectLogger.java]: You are logging the logMessage in requestresponselogger, regardless of if or else, the logMessage is an object where you are appending the string.

File: ida-ws/src/main/java/upb/ida/logging/AspectLogger.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+@Component
+@Aspect
+public class AspectLogger 
+{
+	@Autowired(required=true)
+	private HttpServletRequest request;
+	@Autowired
+	private ResponseBean response;
+	@Autowired
+	private LoggerProvider loggerProvider;
```
Comment (nikit91) [inline ida-ws/src/main/java/upb/ida/logging/AspectLogger.java]: don't autowire the loggerprovider, autowire the logger itself. 
it should look something like this:
```
@Autowired
@Qualifier("requestResponseLogger")
Logger reqRespLogger;
```

File: ida-ws/src/main/java/upb/ida/logging/LoggerProvider.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+	 * Method to get the logger instance for RequestResponseLogger from a log4j.properties file
+	 * 
+	 * @return - RequestResponseLogger
+	 */
+	@Bean
+	@Scope(value = "singleton", proxyMode = ScopedProxyMode.TARGET_CLASS)
+	@Qualifier("requestResponseLogger")
+	public Logger getRequestResponseLogger() 
+	{
+		String log4jConfigFile = fileUtil.fetchSysFilePath("log4j.properties");
```
Comment (nikit91) [inline ida-ws/src/main/java/upb/ida/logging/LoggerProvider.java]: Put the literal "log4j.properties" as a constant  in https://github.com/dice-group/ida/blob/master/ida-ws/src/main/java/upb/ida/constant/IDALiteral.java#L28

File: ida-ws/src/main/java/upb/ida/logging/AspectLogger.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+	@Autowired
+	@Qualifier("datasetResponseLogger")
+	Logger datasetResponseLogger;
+	
+    /**
+     * Constructor of AspectLogger class
+     */
+	AspectLogger()
+	{
+		BasicConfigurator.configure();
```
Comment (nikit91) [inline ida-ws/src/main/java/upb/ida/logging/AspectLogger.java]: could you please confirm once by removing this statement and checking if its really needed?
if yes, then put it in a static block inside your loggerprovider class

File: ida-ws/src/main/java/upb/ida/logging/AspectLogger.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+@Aspect
+public class AspectLogger 
+{
+	@Autowired(required=true)
+	private HttpServletRequest request;
+	@Autowired
+	private ResponseBean response;
+	@Autowired
+	@Qualifier("requestResponseLogger")
+	Logger reqRespLogger;
```
Comment (nikit91) [inline ida-ws/src/main/java/upb/ida/logging/AspectLogger.java]: mark these as private

**Comment by nikit91** [comment]
Hi @RicardoUsbeck ,
I have reviewed the changes made and it all looks good to me.
Could you please have a quick look at this PR as well?

**Comment by nikit91** [comment]
@RicardoUsbeck ,
I requested the sample log files before and you can find them here:
https://github.com/dice-group/ida/pull/79#issuecomment-440768008

**Pull request merged**